src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,300 @@
+
+            let port_mouse_down = on_port_mouse_down.clone_ref().into();
+        }
+        Self{set_size,on_port_mouse_down,port_mouse_down}
+    }
+}
+
+
+
+// =======================
+// === OutPutPortsData ===
+// =======================
+
+#[derive(Debug)]
+pub struct OutPutPortsData {
+    display_object : display::object::Instance,
+    scene          : Scene,
+    logger         : Logger,
+    size           : Vector2<f32>,
+    gap_width      : f32,
+    ports          : Vec<component::ShapeView<port_area::Shape>>,
+}
+
+impl OutPutPortsData {
+    fn init(mut self) -> Self {
+        self.update_shapes();
+        self
+    }
+
+    fn update_ports(&mut self) {
+        let port_num      = self.ports.len() as f32;
+
+        let width         = self.size.x;
+        let height        = self.size.y;
+        let element_width = width / port_num;
+        let element_size  = Vector2::new(element_width,height);
+
+        // Align shapes along width.
+        let x_start = -width/2.0 + NODE_SHAPE_PADDING;
+        let x_delta = element_width;
+        for (index, view) in self.ports.iter().enumerate(){
+            view.display_object().set_parent(&self.display_object);
+
+            let pos_x = x_start + x_delta * index as f32;
+            let pos_y = 0.0;
+            let pos   = Vector2::new(pos_x,pos_y);
+            view.set_position_xy(pos);
+
+            let shape = &view.shape;
+            shape.sprite.size().set(element_size);
+            shape.shape_width.set(width);
+            shape.padding.set(self.gap_width);
+            shape.offset_x.set(x_delta * index as f32);
+        }
+    }
+
+    fn update_shapes(&mut self) {
+        self.update_ports();
+    }
+
+    fn set_size(&mut self, size:Vector2<f32>) {
+        self.size = size;
+        self.update_shapes();
+    }
+}
+
+// ===================
+// === OutPutPorts ===
+// ===================
+
+#[derive(Debug,Clone,CloneRef)]
+pub struct OutputPorts {
+    pub frp            : Frp,
+        network        : frp::Network,
+        data           : Rc<RefCell<OutPutPortsData>>,
+        display_object : display::object::Instance,
+}
+
+impl OutputPorts {
+    pub fn new(scene:&Scene, number_of_ports:u8) -> Self {
+        let logger         = Logger::new("bubble");
+
+        let display_object = display::object::Instance::new(&logger);
+        let network        = default();
+        let frp            = Frp::new(&network);
+
+        let size           = Vector2::zero();
+        let scene          = scene.clone_ref();
+        let gap_width      = SEGMENT_GAP_WIDTH;
+
+        let mut ports      = Vec::default();
+        ports.resize_with(number_of_ports as usize,|| component::ShapeView::new(&logger,&scene));
+        let ports = ports;
+
+        let data =  OutPutPortsData {display_object:display_object.clone_ref(),scene,logger,size,
+                                     ports,gap_width}.init();
+        let data = Rc::new(RefCell::new(data));
+
```

Comment:
A lot of things are badly aligned here. If you need to use such patterns as here, it is always much better to create a `OutPutPortsData::new` method and use it here instead.